client/rpc/src/lib.rs

```diff
@@ -43,3 +43,121 @@ pub mod testing;
 
 /// Task executor that is being used by RPC subscriptions.
 pub type SubscriptionTaskExecutor = std::sync::Arc<dyn sp_core::traits::SpawnNamed>;
+
+/// JSON-RPC helpers.
+pub mod utils {
+	use futures::{channel::oneshot, Future, FutureExt, Stream, StreamExt};
+	use jsonrpsee::{
+		IntoSubscriptionCloseResponse, PendingSubscriptionSink, SendTimeoutError,
+		SubscriptionCloseResponse, SubscriptionMessage, SubscriptionSink,
+	};
+	use sp_runtime::Serialize;
+
+	use crate::SubscriptionTaskExecutor;
+
+	/// Similar to [`pipe_from_stream`] but also attempts to accept the subscription.
+	pub async fn accept_and_pipe_from_stream<R, S, T>(
+		pending: PendingSubscriptionSink,
+		stream: S,
+		executor: &SubscriptionTaskExecutor,
+	) -> SubscriptionResponse<R>
+	where
+		S: Stream<Item = T> + Unpin + Send + 'static,
+		T: Serialize + Send + 'static,
+		R: Serialize + Send + 'static,
+	{
+		let Ok(sink) = pending.accept().await else {
+			return SubscriptionResponse::Closed
+		};
+		pipe_from_stream(sink, stream, executor).await
+	}
+
+	/// Feed items to the subscription from the underlying stream
+	/// if the subscription can't keep up with the underlying stream
+	/// it's dropped
+	///
+	/// This is simply a way to keep previous behaviour with unbounded streams
+	/// and should be replaced by specific RPC endpoint behaviour.
+	pub async fn pipe_from_stream<R, S, T>(
+		sink: SubscriptionSink,
+		mut stream: S,
+		executor: &SubscriptionTaskExecutor,
+	) -> SubscriptionResponse<R>
+	where
+		S: Stream<Item = T> + Unpin + Send + 'static,
+		T: Serialize + Send + 'static,
+		R: Serialize + Send + 'static,
+	{
+		let (tx, rx) = oneshot::channel();
+
+		let fut = async move {
+			let res = loop {
+				tokio::select! {
+					biased;
+					_ = sink.closed() => break SubscriptionResponse::<R>::Closed,
+
+					maybe_item = stream.next() => {
+						let msg = match maybe_item {
+							Some(item) => crate::utils::to_sub_message(&item),
+							None => break SubscriptionResponse::Closed,
+						};
+
+						match sink.send_timeout(msg, std::time::Duration::from_secs(60)).await {
```

Comment:
The major change is here: if the subscription message could not be sent in 60 seconds => drop the subscription

This is really naive and the new spec has specified behavior when to drop etc.. but keep somewhat "backward compatible API".

We could also send a special message when this occurs but the client is most likely disconnected anyway..